python-frontend/src/main/java/org/sonar/python/semantic/v2/types/TrivialTypeInferenceVisitor.java

```diff
@@ -368,13 +402,13 @@ public void visitName(Name name) {
 
     bindingUsages.stream()
       .findFirst()
-      .filter(UsageV2::isBindingUsage)
-      .map(UsageV2::tree)
-      .filter(Expression.class::isInstance)
-      .map(Expression.class::cast)
-      .map(Expression::typeV2)
-      // TODO: classes (SONARPY-1829) and functions should be propagated like other types
-      .filter(t -> (t instanceof ClassType) || (t instanceof FunctionType))
+      .flatMap(usage -> Optional.of(usage)
+        .map(UsageV2::tree)
+        .filter(Expression.class::isInstance)
+        .map(Expression.class::cast)
+        .map(Expression::typeV2)
+        // TODO: classes (SONARPY-1829) and functions should be propagated like other types
+        .filter(t -> (usage.kind() == UsageV2.Kind.PARAMETER) || (t instanceof ClassType) || (t instanceof FunctionType)))
```

Comment:
Should this be here? I have the feeling this piece of code is here to assign types for names referencing classes and functions that are assigned only once, because we cannot propagate them properly.

To me for type hints, the initial propagation should be when the parameter is declared with its type hint, then it should propagate like any other name.

Here, if the parameter is reassigned, such as:
```
def foo(param: int):
    bar(param)
    param = "hello"
    qix(param)
```

We'll not consider that the call to `bar` is made with a declared type `int` because of the check on the number of binding usages above.

Note: I also think we'd need such reassignment scenarios in our tests.